### DISCUSSION TITLE
payment via Stripe (flutter web) ?

### DISCUSSION THREAD

**Discussion by layerzzzio** [serverpod/serverpod#872] [Q&A] [answered]
I am exploring ways to use Stripe in a Flutter web app that uses Serverpod.

Note: I focus on Flutter web apps because when developing apps on Android or iOS, an in-app payment system such as RevenueCat does the job just right.

Back to the initial question. 

FYI I have been implementing payment processes with Stripe before using Firebase and the Stripe extension (from Firebase). It is really easy to set up and use. It essentially uses Google Cloud Functions (webhooks) to create info required for payments and handle cases related to payments. For example, it creates a stripeId and stripeUrl anytime a new user signs up to your app.

Using Serverpod, I was thinking, I could do pretty much the same thing, being: authenticate users using the serverpod_auth_firebase_flutter module and using the Stripe extension as mentioned earlier. Doing so, the payments would be handled by Firebase Auth + Stripe extensions while the rest of the application would be handled by Serverpod.

Any thoughts on that?

**Comment by Isakdl** [comment, marked as answer]
This should be doable only using serverpod, but as far as I know there are no server side libraries for the stripe sdk/api so you would have to build your http requests manually and then talk with stripe via them. 
Basically following the official documentation at: https://stripe.com/docs/api/payment_intents#payment_intents

To manage webhooks you can do that in serverpod using the web server to define callable routes and write your custom logic in these functions. See: https://docs.serverpod.dev/concepts/webserver

And instead of returning a web page you can return an empty WidgetJson to notify stripe the webhook was processed successfully.

For the client side you can of course use the available packages at pub.dev.

**Reply by riccardo-runci** [reply]
I have created a webhook on stripe to a custom route of serverpod web page, but it return 400, I have done something wrong?

this is the code

```
class StripeCheckoutRoot extends WidgetRoute {
  @override
  Future<Widget> build(Session session, HttpRequest request) async {
    request.response.statusCode = HttpStatus.ok;
    request.response.close();
    return StripeCheckoutWidget();
  }
}

class StripeCheckoutWidget extends Widget {
  StripeCheckoutWidget() : super(name: 'stripecheckout') {
    WidgetJson(object: {});
  }
}
```

in `server.dart` I added this line
`  pod.webServer.addRoute(StripeCheckoutRoot(), '/stripe/checkout/success');`

on the stripe Webhook I have added a checkout.session.completed webhook to
`https://myappurl/stripe/checkout/success`

but it return me 400 (bad request)
the server is hosted on google cloud run

**Reply by Isakdl** [reply]
You should not call `close()`.

Calling close here means you are stopping the communication before you have returned your response. Serverpod will do this for you when the request has been processed.
